File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/missingowners/MissingOwnersFragment.java
package com.lucagiorgetti.surprix.ui.mainfragments.missingowners;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.lucagiorgetti.surprix.R;
import com.lucagiorgetti.surprix.model.User;
import com.lucagiorgetti.surprix.utility.BaseFragment;

public class MissingOwnersFragment extends BaseFragment {

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        MissingOwnersViewModel mViewModel = new ViewModelProvider(this).get(MissingOwnersViewModel.class);
        String surpriseId = MissingOwnersFragmentArgs.fromBundle(requireArguments()).getMissingSurpriseId();

        View root = inflater.inflate(R.layout.fragment_missing_owners, container, false);
        View emptyList = root.findViewById(R.id.missing_owner_empty_list);
        RecyclerView recyclerView = root.findViewById(R.id.missing_owner_recycler);
        ProgressBar progress = root.findViewById(R.id.missing_owner_loading);
        setProgressBar(progress);

        SwipeRefreshLayout swipeRefreshLayout = root.findViewById(R.id.swipe_refresh);
        swipeRefreshLayout.setOnRefreshListener(() -> mViewModel.getMissingOwners(surpriseId));

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        MissingOwnersAdapter adapter = new MissingOwnersAdapter();
        adapter.setListener(new MyClickListener(root));

        recyclerView.setAdapter(adapter);

        mViewModel.getMissingOwners(surpriseId).observe(getViewLifecycleOwner(), owners -> {
            emptyList.setVisibility(owners == null || owners.isEmpty() ? View.VISIBLE : View.GONE);
            if (!owners.isEmpty()) {
                adapter.setOwners(owners);
                adapter.notifyDataSetChanged();
            }
        });

        mViewModel.isLoading().observe(getViewLifecycleOwner(), isLoading -> {
            if (isLoading) {
                showLoading();
            } else {
                hideLoading();
                swipeRefreshLayout.setRefreshing(false);
            }
        });

        setTitle(getString(R.string.find_surprise));
        return root;
    }

    static class MyClickListener {
        View root;

        MyClickListener(View v) {
            root = v;
        }

        void onOwnerClicked(User user) {
            Navigation.findNavController(root).navigate(MissingOwnersFragmentDirections.actionNavigationMissingOwnersToNavigationOtherForYou(user.getUsername(), user.getCleanedEmail()));
        }
    }
}

File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/missingowners/MissingOwnersFragment.kt
package com.lucagiorgetti.surprix.ui.mainfragments.missingowners

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.lucagiorgetti.surprix.R
import com.lucagiorgetti.surprix.model.User
import com.lucagiorgetti.surprix.utility.BaseFragment
import com.lucagiorgetti.surprix.utility.SystemUtils

class MissingOwnersFragment : BaseFragment() {
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val mViewModel = ViewModelProvider(this)[MissingOwnersViewModel::class.java]
        val surpriseId = MissingOwnersFragmentArgs.fromBundle(requireArguments()).missingSurpriseId
        val root = inflater.inflate(R.layout.fragment_missing_owners, container, false)
        val emptyList = root.findViewById<View>(R.id.missing_owner_empty_list)
        val recyclerView = root.findViewById<RecyclerView>(R.id.missing_owner_recycler)
        val progress = root.findViewById<ProgressBar>(R.id.missing_owner_loading)
        setProgressBar(progress)
        val swipeRefreshLayout = root.findViewById<SwipeRefreshLayout>(R.id.swipe_refresh)
        swipeRefreshLayout.setOnRefreshListener {
            emptyList.visibility = View.GONE
            mViewModel.loadOwners(surpriseId)
        }
        recyclerView.layoutManager = GridLayoutManager(context, SystemUtils.getColumnsNumber(recyclerView))
        val adapter = MissingOwnersAdapter()
        adapter.setListener(MyClickListener(root))
        recyclerView.adapter = adapter
        mViewModel.getMissingOwners(surpriseId).observe(viewLifecycleOwner) { owners: List<User?>? ->
            emptyList.visibility = if (owners.isNullOrEmpty()) View.VISIBLE else View.GONE
            if (owners!!.isNotEmpty()) {
                adapter.setOwners(owners)
                adapter.notifyDataSetChanged()
            }
        }
        mViewModel.isLoading.observe(viewLifecycleOwner) { isLoading: Boolean ->
            if (isLoading) {
                showLoading()
            } else {
                hideLoading()
                swipeRefreshLayout.isRefreshing = false
            }
        }
        setTitle(getString(R.string.find_surprise))
        return root
    }

    class MyClickListener(var root: View) {
        fun onOwnerClicked(user: User?) {
            findNavController(root).navigate(MissingOwnersFragmentDirections.actionNavigationMissingOwnersToNavigationOtherForYou(user?.username!!, user.clearedEmail()))
        }
    }
}
